### PR TITLE
Prevent data corruption caused by batch id re-use.

### PR DESCRIPTION
**Pull request by rfk**
WIP, for now this is just a failing test...

### CODE REVIEW

**Comment by mostlygeek** [comment]
gosync passes the new functional test.

**Comment by rfk** [comment]
@mostlygeek r?

File: syncstorage/storage/sql/__init__.py [hunk 1]
```diff
@@ -493,7 +493,11 @@ def create_batch(self, session, userid, collection):
             "userid": userid,
             "collection": collectionid
         }
-        session.query("CREATE_BATCH", params)
+        try:
+            session.query("CREATE_BATCH", params)
+        except IntegrityError:
+            # The user tried to create two batches with the same timestamp.
```
Comment (mostlygeek) [inline syncstorage/storage/sql/__init__.py]: will this catch duplicate timestamps from different users ?

Comment (rfk) [inline syncstorage/storage/sql/__init__.py]: no; the primary key on this table is (batchid, userid) so it'll only trigger for duplicates of this tuple.

(This change isn't critical to the issue at hand, but was motivated by the new testcase.  It was possible for a user to trigger a `500 Server Error` by creating a batch and then quickly creating another without closing the first.  Now that will produce a `409 Conflict` which is more appropriate)

**Pull request merged**